File: frontend/src/api/bom.ts
import client from './client'

// ─── Internal form state (per-layer, used inside BomCalculatorPanel) ──────────

export interface BomLayerInput {
  ma_ky_hieu: string | null
  paper_material_id?: number | null
  dinh_luong: number | null
  don_gia_kg: number
}

/** @deprecated Use BomLayerInput instead */
export type BomLayerFormState = BomLayerInput

// ─── API-level layer input (what backend expects in `layers` array) ────────────

export interface BomLayerApiInput {
  vi_tri_lop: string
  loai_lop: 'mat' | 'song'
  flute_type: string | null
  ma_ky_hieu: string
  paper_material_id?: number | null
  dinh_luong: number
  don_gia_kg: number
}

// ─── Request ──────────────────────────────────────────────────────────────────

export interface BomCalculateRequest {
  loai_thung: 'A1' | 'A3' | 'A5' | 'A7' | 'GOI_GIUA' | 'GOI_SUON' | 'TAM'
  dai: number
  rong: number
  cao: number
  so_lop: 3 | 5 | 7
  to_hop_song: string
  layers: BomLayerApiInput[]
  so_luong: number
  // Add-ons
  chong_tham: 0 | 1 | 2
  in_flexo_mau: number
  in_flexo_phu_nen: boolean
  in_ky_thuat_so: boolean
  chap_xa: boolean
  boi: boolean
  be_so_con: 0 | 1 | 2 | 4 | 6 | 8
  can_mang: 0 | 1 | 2
  san_pham_kho: boolean
  // Pricing
  ty_le_loi_nhuan?: number
  hoa_hong_kd_pct: number
  hoa_hong_kh_pct: number
  chi_phi_khac: number
  chiet_khau: number
}

// ─── Response – calculate ─────────────────────────────────────────────────────

export interface BomDimensions {
  kho1: number
  dai1: number
  so_dao: number
  kho_tt: number
  dai_tt: number
  kho_kh: number
  dai_kh: number
  dien_tich: number
}

export interface BomAddonDetail {
  d1_chong_tham: number
  d2_in_flexo: number
  d3_in_ky_thuat_so: number
  d4_chap_xa: number
  d5_boi: number
  d6_be: number
  d8_can_mang: number
  d9_san_pham_kho: number
}

export interface BomLayerResult {
  vi_tri_lop: string
  loai_lop: 'mat' | 'song'
  flute_type: string | null
  ma_ky_hieu: string
  paper_material_id: number | null
  dinh_luong: number
  take_up_factor: number
  dien_tich_1con: number
  trong_luong_1con: number
  chi_phi_1con: number
  so_luong_sx: number
  ty_le_hao_hut: number
  trong_luong_can_tong: number
  don_gia_kg: number
  thanh_tien: number
}

export interface IndirectCostItem {
  ten: string
  don_gia_m2: number
  thanh_tien: number
}

export interface BomCalculateResponse {
  dimensions: BomDimensions
  chi_phi_giay: number
  chi_phi_gian_tiep: number
  ty_le_hao_hut: number
  chi_phi_hao_hut: number
  ty_le_loi_nhuan: number
  loi_nhuan: number
  addon_detail: BomAddonDetail
  chi_phi_addon: number
  gia_ban_co_ban: number
  hoa_hong_kd: number
  hoa_hong_kh: number
  chi_phi_khac: number
  chiet_khau: number
  gia_ban_cuoi: number
  bom_layers: BomLayerResult[]
  gian_tiep_breakdown: IndirectCostItem[]
}

// ─── Response – from-production-item (auto-calculated, no config needed) ────

export interface BomFromProductionItemResponse extends BomCalculateResponse {
  source: 'quote' | 'cau_truc' | 'product'
  loai_thung: string
  dai: number
  rong: number
  cao: number
  so_lop: number
  to_hop_song: string
  so_luong: number
  bien_phi: number
  gia_ban_bao_gia: number
  lai_gop: number
  ty_le_lai: number
}

// ─── Response – saved BOM ─────────────────────────────────────────────────────

export interface BomSavedItem {
  vi_tri_lop: string
  loai_lop: string
  flute_type: string | null
  ma_ky_hieu: string | null
  dinh_luong: number
  take_up_factor: number
  so_luong_sx: number
  ty_le_hao_hut: number
  trong_luong_can_tong: number | null
  don_gia_kg: number
  thanh_tien: number | null
}

export interface BomSavedIndirectItem {
  id: number
  bom_id: number
  ten: string
  don_gia_m2: number
  dien_tich: number
  thanh_tien: number
}

export interface BomSaved {
  id: number
  production_order_item_id: number | null
  loai_thung: string
  dai: number
  rong: number
  cao: number
  so_lop: number
  to_hop_song: string | null
  // Dimensions
  kho_tt: number | null
  dai_tt: number | null
  kho_kh: number | null
  dai_kh: number | null
  dien_tich: number | null
  // Production
  so_luong_sx: number
  ty_le_hao_hut: number | null
  // Costs
  chi_phi_giay: number | null
  chi_phi_gian_tiep: number | null
  chi_phi_hao_hut: number | null
  loi_nhuan: number | null
  chi_phi_addon: number | null
  gia_ban_co_ban: number | null
  gia_ban_cuoi: number | null
  // Add-on config
  chong_tham: number
  in_flexo_mau: number
  in_flexo_phu_nen: boolean
  in_ky_thuat_so: boolean
  chap_xa: boolean
  boi: boolean
  be_so_con: number
  can_mang: number
  san_pham_kho: boolean
  // Pricing
  ty_le_loi_nhuan: number | null
  hoa_hong_kd_pct: number
  hoa_hong_kh_pct: number
  hoa_hong_kd: number | null
  hoa_hong_kh: number | null
  chi_phi_khac: number
  chiet_khau: number
  trang_thai: string
  ghi_chu: string | null
  created_at: string
  updated_at: string
  items: BomSavedItem[]
  indirect_items: BomSavedIndirectItem[]
}

export interface BomSaveResponse {
  bom_id: number
  message: string
}

// ─── API Client ───────────────────────────────────────────────────────────────

// ─── Reverse calculation ──────────────────────────────────────────────────────

export interface BomReverseRequest {
  gia_muc_tieu: number
  loai_thung: string
  dai: number
  rong: number
  cao: number
  so_lop: 3 | 5 | 7
  so_luong?: number
  ty_le_loi_nhuan?: number
  d_total?: number
  hoa_hong_kd_pct?: number
  hoa_hong_kh_pct?: number
  chi_phi_khac?: number
  chiet_khau?: number
}

export interface BomReverseResponse {
  gia_muc_tieu: number
  p_co_ban: number
  b_per_m2: number
  b: number
  c_pct: number
  e_pct: number
  d: number
  a_max: number
  a_max_per_m2: number
  dien_tich: number
  kha_thi: boolean
}

// ─── Indirect cost items (master data) ───────────────────────────────────────

export interface IndirectCostMasterItem {
  id: number
  so_lop: number
  ten: string
  don_gia_m2: number
  thu_tu: number
  ghi_chu: string | null
}

export const indirectCostsApi = {
  list: () => client.get<IndirectCostMasterItem[]>('/indirect-costs'),
  update: (id: number, data: { ten?: string; don_gia_m2?: number; thu_tu?: number }) =>
    client.put<IndirectCostMasterItem>(`/indirect-costs/${id}`, data),
  seed: () => client.post('/indirect-costs/seed', {}),
}

export interface AddonRateItem {
  id: number
  ma_chi_phi: string
  nhom: string
  ten: string
  don_vi: string  // 'm2' | 'pcs' | 'pct'
  don_gia: number
  ghi_chu: string | null
  thu_tu: number
}

export const addonRatesApi = {
  list: () => client.get<AddonRateItem[]>('/addon-rates'),
  update: (id: number, data: { ten?: string; don_gia?: number; ghi_chu?: string }) =>
    client.put<AddonRateItem>(`/addon-rates/${id}`, data),
  seed: () => client.post('/addon-rates/seed', {}),
}

// ─── API Client ───────────────────────────────────────────────────────────────

// Quy cách sản phẩm từ báo giá (để auto-fill BOM calculator)
export interface QuoteSpec {
  source: 'quote' | 'cau_truc' | 'product'
  quote_item_id: number | null
  loai_thung: string
  dai: number | null
  rong: number | null
  cao: number | null
  so_lop: number
  to_hop_song: string
  so_luong: number
  layers: BomLayerApiInput[]
  chong_tham: number
  in_flexo_mau: number
  in_flexo_phu_nen: boolean
  in_ky_thuat_so: boolean
  chap_xa: boolean
  boi: boolean
  be_so_con: number
  can_mang: number
  san_pham_kho: boolean
}

// ─── BOM Summary (for list page) ─────────────────────────────────────────────

export interface BomSummaryItem {
  id: number
  production_order_item_id: number | null
  ten_hang: string | null
  so_lenh: string | null
  ten_khach_hang: string | null
  ma_khach_hang: string | null
  loai_thung: string
  dai: number
  rong: number
  cao: number
  so_lop: number
  to_hop_song: string | null
  so_luong_sx: number
  chi_phi_giay: number | null
  chi_phi_gian_tiep: number | null
  chi_phi_hao_hut: number | null
  chi_phi_addon: number | null
  gia_ban_cuoi: number | null
  trang_thai: string
  created_at: string
  updated_at: string
}

export const bomApi = {
  listSummary: (params?: { trang_thai?: string; search?: string }) =>
    client.get<BomSummaryItem[]>('/bom/summary', { params }),

  calculate: (request: BomCalculateRequest) =>
    client.post<BomCalculateResponse>('/bom/calculate', request),

  save: (request: BomCalculateRequest & { production_order_item_id?: number; ghi_chu?: string }) =>
    client.post<BomSaved>('/bom/save', request),

  confirm: (bomId: number) =>
    client.patch<BomSaved>(`/bom/${bomId}/confirm`),

  getByItem: (production_order_item_id: number) =>
    client.get<BomSaved>(`/bom/by-item/${production_order_item_id}`),

  get: (bomId: number) => client.get<BomSaved>('/bom/' + bomId),

  getQuoteSpec: (production_order_item_id: number) =>
    client.get<QuoteSpec>(`/bom/quote-spec/${production_order_item_id}`),

  reverseCalculate: (request: BomReverseRequest) =>
    client.post<BomReverseResponse>('/bom/reverse-calculate', request),

  fromProductionItem: (productionOrderItemId: number, soLuong?: number) =>
    client.get<BomFromProductionItemResponse>(
      `/bom/from-production-item/${productionOrderItemId}`,
      soLuong !== undefined ? { params: { so_luong: soLuong } } : undefined,
    ),
}

// ─── Display helpers ──────────────────────────────────────────────────────────

export const LOAI_THUNG_BOM_OPTIONS: Array<{ value: string; label: string }> = [
  { value: 'A1',       label: 'A1 – Thùng thường' },
  { value: 'A3',       label: 'A3 – Nắp chồm' },
  { value: 'A5',       label: 'A5 – Âm dương (Nắp/Đáy)' },
  { value: 'A7',       label: 'A7 – Thùng 1 nắp' },
  { value: 'GOI_GIUA', label: 'Gói giữa' },
  { value: 'GOI_SUON', label: 'Gói sườn' },
  { value: 'TAM',      label: 'Giấy tấm' },
]

export const SO_LOP_BOM_OPTIONS: Array<{ value: number; label: string }> = [
  { value: 3, label: '3 lớp' },
  { value: 5, label: '5 lớp' },
  { value: 7, label: '7 lớp' },
]

// Tổ hợp sóng suggestions by so_lop
export const TO_HOP_SONG_BY_LOP: Record<number, string[]> = {
  3: ['C', 'B', 'E', 'A'],
  5: ['CB', 'CC', 'BC', 'BE', 'AB', 'CE'],
  7: ['CBC', 'BCB', 'CBE', 'ABE', 'BCE'],
}

export const BE_SO_CON_OPTIONS: Array<{ value: number; label: string }> = [
  { value: 0, label: 'Không bế' },
  { value: 1, label: '1 con' },
  { value: 2, label: '2 con' },
  { value: 4, label: '4 con' },
  { value: 6, label: '6 con' },
  { value: 8, label: '8 con' },
]

export const CHONG_THAM_OPTIONS: Array<{ value: number; label: string }> = [
  { value: 0, label: 'Không' },
  { value: 1, label: '1 mặt' },
  { value: 2, label: '2 mặt' },
]

export const CAN_MANG_OPTIONS: Array<{ value: number; label: string }> = [
  { value: 0, label: 'Không' },
  { value: 1, label: '1 mặt' },
  { value: 2, label: '2 mặt' },
]

// Vietnamese currency formatter
export const vnd = (value: number) =>
  new Intl.NumberFormat('vi-VN').format(Math.round(value))

File: backend/app/schemas/bom.py
from datetime import datetime
from decimal import Decimal
from typing import Any
from pydantic import BaseModel, Field, field_validator


# ---------------------------------------------------------------------------
# Layer input schema (dùng trong request)
# ---------------------------------------------------------------------------

class BomLayerInput(BaseModel):
    """Thông tin một lớp giấy khi gửi lên để tính BOM."""
    vi_tri_lop: str = Field(..., description="Tên vị trí lớp, vd: 'Mặt ngoài', 'Sóng C'")
    loai_lop: str = Field(..., description="'mat' hoặc 'song'")
    flute_type: str | None = Field(None, description="E/B/C/A — chỉ cho lớp sóng")
    ma_ky_hieu: str = Field(..., description="Mã ký hiệu giấy")
    paper_material_id: int | None = Field(None, description="ID nguyên liệu giấy (nếu đã có)")
    dinh_luong: Decimal = Field(..., description="Định lượng g/m²", gt=0)
    don_gia_kg: Decimal = Field(..., description="Đơn giá mua đ/kg", ge=0)

    @field_validator("loai_lop")
    @classmethod
    def validate_loai_lop(cls, v: str) -> str:
        if v not in ("mat", "song"):
            raise ValueError("loai_lop phải là 'mat' hoặc 'song'")
        return v


# ---------------------------------------------------------------------------
# BomCalculateRequest
# ---------------------------------------------------------------------------

class BomCalculateRequest(BaseModel):
    """Dữ liệu đầu vào để tính BOM + giá."""

    # Thông số sản phẩm
    loai_thung: str = Field(..., description="A1 | A3 | A5 | tam")
    dai: Decimal = Field(..., description="Chiều dài (cm)", gt=0)
    rong: Decimal = Field(..., description="Chiều rộng (cm)", gt=0)
    cao: Decimal = Field(..., description="Chiều cao (cm)", ge=0)
    so_lop: int = Field(..., description="Số lớp: 3, 5 hoặc 7")
    to_hop_song: str = Field(..., description="Tổ hợp sóng, vd: 'C-B', 'B', 'BCE'")
    so_luong: Decimal = Field(..., description="Số lượng sản xuất", gt=0)

    # Các lớp giấy (theo thứ tự từ ngoài vào trong)
    layers: list[BomLayerInput] = Field(..., min_length=3)

    # Add-ons
    chong_tham: int = Field(0, description="0=không, 1=1 mặt, 2=2 mặt", ge=0, le=2)
    in_flexo_mau: int = Field(0, description="0=không in, 1+=số màu", ge=0)
    in_flexo_phu_nen: bool = Field(False)
    in_ky_thuat_so: bool = Field(False)
    chap_xa: bool = Field(False)
    boi: bool = Field(False)
    be_so_con: int = Field(0, description="0/1/2/4/6/8")
    dan: bool = Field(False, description="Dán thùng")
    ghim: bool = Field(False, description="Ghim thùng")
    can_mang: int = Field(0, description="0=không, 1=1 mặt, 2=2 mặt", ge=0, le=2)
    san_pham_kho: bool = Field(False)

    # Định giá
    ty_le_loi_nhuan: Decimal | None = Field(
        None, description="Override tỷ lệ lợi nhuận. None = dùng mặc định theo loại+số lớp"
    )
    hoa_hong_kd_pct: Decimal = Field(Decimal("0"), description="% hoa hồng KD", ge=0)
    hoa_hong_kh_pct: Decimal = Field(Decimal("0"), description="% hoa hồng KH", ge=0)
    chi_phi_khac: Decimal = Field(Decimal("0"), description="Chi phí khác (đồng)", ge=0)
    chiet_khau: Decimal = Field(Decimal("0"), description="Chiết khấu (đồng)", ge=0)

    @field_validator("so_lop")
    @classmethod
    def validate_so_lop(cls, v: int) -> int:
        if v not in (3, 5, 7):
            raise ValueError("so_lop phải là 3, 5 hoặc 7")
        return v

    @field_validator("loai_thung")
    @classmethod
    def validate_loai_thung(cls, v: str) -> str:
        valid = {"A1", "A3", "A5", "A7", "GOI_GIUA", "GOI_SUON", "TAM"}
        if v.upper() not in valid:
            raise ValueError(f"loai_thung phải là {', '.join(sorted(valid))}")
        return v.upper()

    @field_validator("be_so_con")
    @classmethod
    def validate_be_so_con(cls, v: int) -> int:
        if v not in (0, 1, 2, 4, 6, 8):
            raise ValueError("be_so_con phải là 0, 1, 2, 4, 6 hoặc 8")
        return v


# ---------------------------------------------------------------------------
# BomSaveRequest (extends calculate with FK)
# ---------------------------------------------------------------------------

class BomSaveRequest(BomCalculateRequest):
    """Tính BOM và lưu vào DB."""
    production_order_item_id: int | None = Field(
        None, description="ID dòng lệnh SX liên kết (tuỳ chọn)"
    )
    ghi_chu: str | None = None


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------

class DimensionResult(BaseModel):
    kho1: float
    dai1: float
    so_dao: int
    kho_tt: float
    dai_tt: float
    kho_kh: float
    dai_kh: float
    dien_tich: float  # m²/unit


class AddonDetail(BaseModel):
    d1_chong_tham: float = 0
    d2_in_flexo: float = 0
    d3_in_ky_thuat_so: float = 0
    d4_chap_xa: float = 0
    d5_boi: float = 0
    d6_be: float = 0
    d7_dan: float = 0
    d7_ghim: float = 0
    d8_can_mang: float = 0
    d9_san_pham_kho: float = 0


class BomLayerResult(BaseModel):
    """Kết quả tính toán cho một lớp giấy."""
    vi_tri_lop: str
    loai_lop: str
    flute_type: str | None = None
    ma_ky_hieu: str
    paper_material_id: int | None = None
    dinh_luong: float
    take_up_factor: float
    dien_tich_1con: float    # m²/unit (after take-up for song layers)
    trong_luong_1con: float  # kg/unit
    don_gia_kg: float
    chi_phi_1con: float
    so_luong_sx: float
    ty_le_hao_hut: float
    trong_luong_can_tong: float
    thanh_tien: float


class IndirectCostItem(BaseModel):
    ten: str
    don_gia_m2: float
    thanh_tien: float


class BomCalculateResponse(BaseModel):
    """Full response từ endpoint /calculate."""
    dimensions: DimensionResult
    chi_phi_giay: float
    chi_phi_gian_tiep: float
    ty_le_hao_hut: float
    chi_phi_hao_hut: float
    ty_le_loi_nhuan: float
    loi_nhuan: float
    addon_detail: AddonDetail
    chi_phi_addon: float
    gia_ban_co_ban: float
    hoa_hong_kd: float
    hoa_hong_kh: float
    chi_phi_khac: float
    chiet_khau: float
    gia_ban_cuoi: float
    bom_layers: list[BomLayerResult]
    gian_tiep_breakdown: list[IndirectCostItem] = []


# ---------------------------------------------------------------------------
# Saved BOM response schemas
# ---------------------------------------------------------------------------

class BomItemResponse(BaseModel):
    id: int
    bom_id: int
    vi_tri_lop: str
    loai_lop: str
    flute_type: str | None
    ma_ky_hieu: str | None
    paper_material_id: int | None
    dinh_luong: Decimal
    take_up_factor: Decimal
    dien_tich_1con: Decimal | None
    trong_luong_1con: Decimal | None
    so_luong_sx: Decimal
    ty_le_hao_hut: Decimal
    trong_luong_can_tong: Decimal | None
    don_gia_kg: Decimal
    thanh_tien: Decimal | None

    class Config:
        from_attributes = True


class BomResponse(BaseModel):
    id: int
    production_order_item_id: int | None

    # Product specs
    loai_thung: str
    dai: Decimal
    rong: Decimal
    cao: Decimal
    so_lop: int
    to_hop_song: str | None

    # Dimensions
    kho_tt: Decimal | None
    dai_tt: Decimal | None
    kho_kh: Decimal | None
    dai_kh: Decimal | None
    dien_tich: Decimal | None

    # Production params
    so_luong_sx: Decimal
    ty_le_hao_hut: Decimal | None

    # Cost breakdown
    chi_phi_giay: Decimal | None
    chi_phi_gian_tiep: Decimal | None
    chi_phi_hao_hut: Decimal | None
    loi_nhuan: Decimal | None
    chi_phi_addon: Decimal | None
    gia_ban_co_ban: Decimal | None
    gia_ban_cuoi: Decimal | None

    # Add-on config
    chong_tham: int
    in_flexo_mau: int
    in_flexo_phu_nen: bool
    in_ky_thuat_so: bool
    chap_xa: bool
    boi: bool
    be_so_con: int
    dan: bool = False
    ghim: bool = False
    can_mang: int
    san_pham_kho: bool

    # Pricing
    ty_le_loi_nhuan: Decimal | None
    hoa_hong_kd_pct: Decimal
    hoa_hong_kh_pct: Decimal
    chi_phi_khac: Decimal
    chiet_khau: Decimal
    hoa_hong_kd: Decimal | None
    hoa_hong_kh: Decimal | None

    # Meta
    trang_thai: str
    ghi_chu: str | None
    created_by: int | None
    created_at: datetime
    updated_at: datetime

    # BOM lines
    items: list[BomItemResponse] = []

    # Chi tiết chi phí gián tiếp (hoạch toán)
    indirect_items: list["BomIndirectItemResponse"] = []

    class Config:
        from_attributes = True


class BomIndirectItemResponse(BaseModel):
    id: int
    bom_id: int
    ten: str
    don_gia_m2: Decimal
    dien_tich: Decimal
    thanh_tien: Decimal


class BomSummaryItem(BaseModel):
    """Lightweight summary dùng cho trang danh sách Định mức BOM."""
    id: int
    production_order_item_id: int | None
    # Context từ quan hệ
    ten_hang: str | None
    so_lenh: str | None
    ten_khach_hang: str | None
    ma_khach_hang: str | None
    # Thông số
    loai_thung: str
    dai: Decimal
    rong: Decimal
    cao: Decimal
    so_lop: int
    to_hop_song: str | None
    so_luong_sx: Decimal
    # Chi phí tổng hợp
    chi_phi_giay: Decimal | None
    chi_phi_gian_tiep: Decimal | None
    chi_phi_hao_hut: Decimal | None
    chi_phi_addon: Decimal | None
    gia_ban_cuoi: Decimal | None
    # Trạng thái
    trang_thai: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

    class Config:
        from_attributes = True
